packages/react-components/react-utilities/src/hooks/useMotionPresence.ts

```diff
@@ -0,0 +1,355 @@
+import * as React from 'react';
+import { canUseDOM } from '../ssr/canUseDOM';
+import { useAnimationFrame } from './useAnimationFrame';
+import { useTimeout } from './useTimeout';
+
+/**
+ * CSS Typed Object Model
+ * @see https://drafts.css-houdini.org/css-typed-om-1/
+ * @see https://developer.mozilla.org/en-US/docs/Web/API/CSSUnitValue
+ */
+interface CSSUnitValue {
+  value: number;
+  readonly unit: string;
+}
+
+/**
+ * Style property map read only.
+ * @see https://developer.mozilla.org/en-US/docs/Web/API/StylePropertyMapReadOnly
+ */
+interface StylePropertyMapReadOnly {
+  [Symbol.iterator](): IterableIterator<[string, CSSUnitValue[]]>;
+
+  get(property: string): CSSUnitValue | undefined;
+  getAll(property: string): CSSUnitValue[];
+  has(property: string): boolean;
+  readonly size: number;
+}
+
+/**
+ * HTMLElement with styled map.
+ * @see https://developer.mozilla.org/en-US/docs/Web/API/Element/computedStyleMap
+ */
+type HTMLElementWithStyledMap<TElement extends HTMLElement = HTMLElement> = TElement & {
+  computedStyleMap(): StylePropertyMapReadOnly;
+};
+
+type CSSWithNumber = typeof CSS & {
+  number(value: number): {
+    value: number;
+    readonly unit: string;
+  };
+};
+
+/**
+ * @internal
+ *
+ * State for useMotionPresence hook.
+ */
+export type UseMotionPresenceState<TElement extends HTMLElement> = {
+  /**
+   * Ref to the element.
+   */
+  ref: React.RefCallback<TElement>;
+
+  /**
+   * Whether the element should be rendered in the DOM.
+   * This should be used to conditionally render the element.
+   */
+  shouldRender: boolean;
+
+  /**
+   * Whether the element is currently visible in the DOM.
+   */
+  visible: boolean;
+
+  /**
+   * Current state of the element.
+   *
+   * - `entering` - The element is entering the DOM.
+   * - `exiting` - The element is exiting the DOM.
+   * - `resting` - The element is currently not animating. This is the final and initial state of the element.
+   * - `unmounted` - The element is not rendered in the DOM.
+   */
+  motionState: 'entering' | 'exiting' | 'resting' | 'unmounted';
```

Comment:
I've been trying to use this hook in https://github.com/microsoft/fluentui/pull/28780

The main problem that I have currently is that there is no `exited` state. I cannot use the `unmounted` state because there is no way in an effect that I can tell the difference between the initial `unmounted` phase and the `unmounted` phase after exist.

Reply:
I could use a ref and modify the effect to tell te differnce between the initial `unmounted` phase. But then I find that the timing is out of sync, so there are flashes as the toasts reappear after the exit animation is finished


```ts
  const firstMountedRef = React.useRef(true);
  useIsomorphicLayoutEffect(() => {
    if (motionState !== 'exiting') {
      return;
    }

    if (firstUnmountedRef.current) {
      firstUnmountedRef.current = false;
      return;
    }

    remove();
  }, [motionState, remove]);
```